tartufo/scanner.py

```diff
@@ -90,35 +91,269 @@ def file_path(self) -> Optional[str]:
     @property
     def signature(self) -> Optional[str]:
         if self.file_path:
-            return generate_signature(self.matched_string, self.file_path)
+            return util.generate_signature(self.matched_string, self.file_path)
         return None
 
     def __str__(self) -> str:
         output = []
         diff_body = self.printable_diff
         diff_body = diff_body.replace(
-            self.matched_string, style_warning(self.matched_string)
+            self.matched_string, util.style_warning(self.matched_string)
         )
         output.append(self.OUTPUT_SEPARATOR)
-        output.append(style_ok("Reason: {}".format(self.issue_type.value)))  # type: ignore
+        output.append(util.style_ok("Reason: {}".format(self.issue_type.value)))  # type: ignore
         if self.issue_detail:
-            output.append(style_ok("Detail: {}".format(self.issue_detail)))
+            output.append(util.style_ok("Detail: {}".format(self.issue_detail)))
         if self.file_path:
-            output.append(style_ok("Filepath: {}".format(self.file_path)))
+            output.append(util.style_ok("Filepath: {}".format(self.file_path)))
         if self.signature:
-            output.append(style_ok("Signature: {}".format(self.signature)))
+            output.append(util.style_ok("Signature: {}".format(self.signature)))
         if self.branch_name:
-            output.append(style_ok("Branch: {}".format(self.branch_name)))
+            output.append(util.style_ok("Branch: {}".format(self.branch_name)))
         if self.commit:
-            output.append(style_ok("Date: {}".format(self.commit_time)))
-            output.append(style_ok("Hash: {}".format(self.commit_hash)))
-            output.append(style_ok("Commit: {}".format(self.commit_message)))
+            output.append(util.style_ok("Date: {}".format(self.commit_time)))
+            output.append(util.style_ok("Hash: {}".format(self.commit_hash)))
+            output.append(util.style_ok("Commit: {}".format(self.commit_message)))
 
         output.append(diff_body)
         output.append(self.OUTPUT_SEPARATOR)
         return "\n".join(output)
 
 
+class ScannerBase(abc.ABC):
+    _issues: Optional[List[Issue]] = None
+    _included_paths: Optional[List[Pattern]] = None
+    _excluded_paths: Optional[List[Pattern]] = None
+    _rules_regexes: Optional[Dict[str, Pattern]] = None
+    options: GlobalOptions
+
+    def __init__(self, options: GlobalOptions) -> None:
+        self.options = options
+
+    @property
+    def issues(self) -> List[Issue]:
+        if self._issues is None:
+            self._issues = self.scan()
+        return self._issues
```

Comment:
The inspiration behind this is the way that forms work in Django. Specifically, when a user submits data to a form, you can have code along the lines of:

```python
form = MyForm(data=request.POST)
if form.errors:
    # Handle errors
```
In accessing `form.errors`, it will automatically "clean" the data for you by calling `self.clean_all()`. This also provides a layer of caching, so that it doesn't have to re-clean the input on subsequent calls.

My thought was to provide something similar in here, so that you could have consuming code along the lines of:

```python
scanner = MyScanner("/some/path")
if scanner.issues:
    # Handle issues
```
As @bosatsu mentioned, this may not end up staying here. I'm pretty iffy on whether this is a desirable usage pattern. And as these classes start getting implemented, a "proper" usage pattern should start to shake out. Any input and opinions are also very welcome here.